gpflow/expectations.py

```diff
@@ -364,57 +361,108 @@ def _expectation(p, kern1, feat1, kern2, feat2, nghp=None):
+
         C_inv_mu = tf.matrix_triangular_solve(C, tf.expand_dims(Xmu, 2), lower=True)  # NxDx1
-        C_inv_z = tf.matrix_triangular_solve(C,
-                                             tf.tile(tf.expand_dims(tf.transpose(Z) / 2., 0), [N, 1, 1]),
-                                             lower=True)  # NxDxM
         mu_CC_inv_mu = tf.expand_dims(tf.reduce_sum(tf.square(C_inv_mu), 1), 2)  # Nx1x1
-        z_CC_inv_z = tf.reduce_sum(tf.square(C_inv_z), 1)  # NxM
-        zm_CC_inv_zn = tf.matmul(C_inv_z, C_inv_z, transpose_a=True)  # NxMxM
-        two_z_CC_inv_mu = 2 * tf.matmul(C_inv_z, C_inv_mu, transpose_a=True)[:, :, 0]  # NxM
-
-        exponent_mahalanobis = mu_CC_inv_mu + tf.expand_dims(z_CC_inv_z, 1) + \
-                               tf.expand_dims(z_CC_inv_z, 2) + 2 * zm_CC_inv_zn - \
-                               tf.expand_dims(two_z_CC_inv_mu, 2) - tf.expand_dims(two_z_CC_inv_mu, 1)  # NxMxM
-        exponent_mahalanobis = tf.exp(-0.5 * exponent_mahalanobis)  # NxMxM
-
-        # Compute sqrt(self.K(Z)) explicitly to prevent automatic gradient from
-        # being NaN sometimes, see pull request #615
-        kernel_sqrt = tf.exp(-0.25 * kern.scaled_square_dist(Z, None))
-        return kern.variance ** 2 * kernel_sqrt * \
+
+        C_inv_z1, z1_CC_inv_z1 = get_cholesky_solve_terms(Z1 / La * half_mean_L)
+        z1_CC_inv_mu = 2 * tf.matmul(C_inv_z1, C_inv_mu, transpose_a=True)[:, :, 0]  # NxM1
+
+        if feat1 == feat2 and Ka == Kb:
+            # in this case Z2==Z1 so we  can reuse the Z1 terms
+            C_inv_z2, z2_CC_inv_z2  = C_inv_z1, z1_CC_inv_z1
+            z2_CC_inv_mu = z1_CC_inv_mu  # NxM
+            Z2 = Z1
+        else:
+            # compute terms related to Z2
+            Z2, _ = Kb._slice(feat2.Z, p.mu)
+            C_inv_z2, z2_CC_inv_z2 = get_cholesky_solve_terms(Z2 / Lb * half_mean_L)
+            z2_CC_inv_mu = 2 * tf.matmul(C_inv_z2, C_inv_mu, transpose_a=True)[:, :, 0]  # NxM2
+
+        z1_CC_inv_z2 = tf.matmul(C_inv_z1, C_inv_z2, transpose_a=True)  # NxM1xM2
+
+        # expand dims for broadcasting
+        # along M1
+        z2_CC_inv_z2 = tf.expand_dims(z2_CC_inv_z2, 1)
+        z2_CC_inv_mu = tf.expand_dims(z2_CC_inv_mu, 1)
+
+        # along M2
+        z1_CC_inv_mu = tf.expand_dims(z1_CC_inv_mu, 2)
+        z1_CC_inv_z1 = tf.expand_dims(z1_CC_inv_z1, 2)
+
+        # expanded version of ((Z1 + Z2)-mu) (CCT)-1 ((Z1 + Z2)-mu)
+        exponent_mahalanobis = mu_CC_inv_mu + z2_CC_inv_z2 + \
+                               z1_CC_inv_z1 + 2 * z1_CC_inv_z2 - \
+                               z1_CC_inv_mu- z2_CC_inv_mu  # NxM1xM2
+
+        exponent_mahalanobis = tf.exp(-0.5 * exponent_mahalanobis)  # NxM1xM2
+
+        if Z1 == Z2:
+            # CAVEAT : Compute sqrt(self.K(Z)) explicitly
+            # to prevent automatic gradient from
+            # being NaN sometimes, see pull request #615
+            dist_term = tf.exp(-0.25 * Ka.scaled_square_dist(Z1, None))
```

Comment:
Even if `Z1 == Z2`, shouldn't the lengthscales here be `lengthscales_rms` like in the `else` branch? Hold on, I'm going to build a failing test case.

Reply:
I don't think so, as this Z1 and Z2 are not same as feat1.Z and feat2.Z they are already scaled. Moreover, unless the objects are not same this branch is never visited. This condition was added to make sure we do not lose performance for both rbfs being exactly same.

Reply:
I agree with @sanket-kamthe here. `Ka.scaled_square_dist` takes care of the square rooting and scaling.